File: frontend/admin/src/api/scheduler.test.ts
import { describe, expect, it, vi } from 'vitest';
import { loadSchedulerConfig, saveSchedulerConfig } from './scheduler';

describe('scheduler api client', () => {
  it('adds admin token header when loading config', async () => {
    const fetchMock = vi.fn().mockResolvedValue({
      ok: true,
      json: async () => ({ enabled: false, mode: 'interval' })
    });
    vi.stubGlobal('fetch', fetchMock);

    await loadSchedulerConfig('secret-token');

    expect(fetchMock).toHaveBeenCalledWith('/admin/scheduler/config', {
      headers: { Authorization: 'Bearer secret-token' }
    });
  });

  it('sends scheduler config as json', async () => {
    const fetchMock = vi.fn().mockResolvedValue({
      ok: true,
      json: async () => ({ enabled: true, mode: 'interval' })
    });
    vi.stubGlobal('fetch', fetchMock);

    await saveSchedulerConfig('secret-token', {
      enabled: true,
      mode: 'interval',
      interval_minutes: 60,
      fixed_times: [],
      concurrency: 2,
      batch_size: 20,
      timeout_seconds: 180,
      source_filter: {
        provider_key: 'llm_web_research',
        ingest_channel: 'ai_web_research',
        source_type: 'news'
      },
      timezone: 'Asia/Shanghai'
    });

    expect(fetchMock).toHaveBeenCalledWith('/admin/scheduler/config', expect.objectContaining({
      method: 'PUT',
      headers: {
        Authorization: 'Bearer secret-token',
        'Content-Type': 'application/json'
      }
    }));
  });

  it('includes api error message when request fails', async () => {
    const fetchMock = vi.fn().mockResolvedValue({
      ok: false,
      status: 401,
      json: async () => ({ error: 'unauthorized' })
    });
    vi.stubGlobal('fetch', fetchMock);

    await expect(loadSchedulerConfig('wrong-token')).rejects.toThrow('unauthorized');
  });
});

File: backend/internal/repositories/scheduler_test.go
package repositories

import (
	"context"
	"reflect"
	"testing"

	"github.com/meierlink88/tidewise-ai/backend/internal/domain"
)

func TestInMemoryRepositoryLoadsDefaultSchedulerConfig(t *testing.T) {
	repo := NewInMemoryRepository(nil)

	config, err := repo.LoadSchedulerConfig(context.Background())
	if err != nil {
		t.Fatalf("LoadSchedulerConfig() error = %v", err)
	}

	if config.Enabled {
		t.Fatal("default scheduler config must be disabled")
	}
	if config.Mode != domain.SchedulerModeInterval {
		t.Fatalf("Mode = %q, want %q", config.Mode, domain.SchedulerModeInterval)
	}
	if config.IntervalMinutes != 60 {
		t.Fatalf("IntervalMinutes = %d, want 60", config.IntervalMinutes)
	}
	if config.SourceFilter.ProviderKey != "" || config.SourceFilter.IngestChannel != "" || config.SourceFilter.SourceType != "" {
		t.Fatalf("SourceFilter = %+v, want empty global filter", config.SourceFilter)
	}
}

func TestInMemoryRepositorySavesSchedulerConfig(t *testing.T) {
	repo := NewInMemoryRepository(nil)
	config := domain.SchedulerConfig{
		ID:             "default",
		Enabled:        true,
		Mode:           domain.SchedulerModeFixedTimes,
		FixedTimes:     []string{"09:00", "12:00", "15:00", "18:00", "21:00"},
		Concurrency:    3,
		BatchSize:      30,
		TimeoutSeconds: 240,
		Timezone:       "Asia/Shanghai",
		SourceFilter: domain.SchedulerSourceFilter{
			ProviderKey:   "llm_web_research",
			IngestChannel: "ai_web_research",
			SourceType:    "news",
		},
	}

	saved, err := repo.SaveSchedulerConfig(context.Background(), config)
	if err != nil {
		t.Fatalf("SaveSchedulerConfig() error = %v", err)
	}
	loaded, err := repo.LoadSchedulerConfig(context.Background())
	if err != nil {
		t.Fatalf("LoadSchedulerConfig() error = %v", err)
	}

	if saved.ConfigVersion != 1 {
		t.Fatalf("saved ConfigVersion = %d, want 1", saved.ConfigVersion)
	}
	if !reflect.DeepEqual(loaded.FixedTimes, config.FixedTimes) {
		t.Fatalf("FixedTimes = %v, want %v", loaded.FixedTimes, config.FixedTimes)
	}
	if loaded.SourceFilter.ProviderKey != "llm_web_research" {
		t.Fatalf("ProviderKey = %q, want llm_web_research", loaded.SourceFilter.ProviderKey)
	}
}
